Allow AsTable on RHS in transformations

pdeffebach opened the pull request:
After the recent improvements to `AsTable` in DataFrames, I guess it's time to make it easier to work with this construct in DataFramesMeta.jl. 

No tests added yet.

bkamins commented:
What would be the test cases for this?

pdeffebach commented:
a row-wise sum

```
cols = [:a, :b, :c, :d]
@rtransform df :y = sum(AsTable(cols))
```

bkamins commented:
Nice - probably documentation should be updated (or does it already have it?)

pdeffebach commented:
No I haven't done any of that yet. I'll ping you when its ready.

pdeffebach commented:
Okay I've added enough stuff so that you can see the API

```

julia> df = DataFrame(rand(10, 10), :auto);

julia> @rselect df :sum_all = sum(AsTable(names(df)))
10×1 DataFrame
 Row │ sum_all 
     │ Float64 
─────┼─────────
   1 │ 4.19633
   2 │ 4.69652
   3 │ 2.96451
   4 │ 4.32973
   5 │ 4.69788
   6 │ 3.68512
   7 │ 6.47631
   8 │ 5.80387
   9 │ 4.92043
  10 │ 4.83637
```

pdeffebach commented:
Okay ready for a review @bkamins @nalimilan 

If only to check if the API makes sense.

bkamins commented on src/macros.jl:
```diff
@@ -630,6 +630,39 @@ macro with(d, body)
+In transformations, it is also allowed to use `AsTable(cols)` to work with
+multiple columns at once, where the columns are grouped together in a
+`NamedTuple`. When `AsTable(cols)` appears in a transformation, no
+other columns may be referenced in the block.
```
how are then symbols outside of `AsTable` treated - they error or are treated as symbols?

bkamins replied:
Also can `AsTable` appear in the specification multiple times?

Also is something like `target = AsTable(nms)` valid? And what should/can `target` be then?

pdeffebach replied:
Good questions. 

The error would be the same as `transform(df, [AsTable([:a, :b]), :c] => ...)`

```
julia> @select df :y = begin 
           nothing
           AsTable([:x1, :x2])
       end
ERROR: ArgumentError: Table returned but a single output column was expected
Stacktrace:
```

but the following works

```
julia> @select df $AsTable = begin 
           nothing
           AsTable([:x1, :x2])
       end
10×2 DataFrame
 Row │ x1         x2       
     │ Float64    Float64  
─────┼─────────────────────
   1 │ 0.383283   0.394874
   2 │ 0.173797   0.285121
   3 │ 0.0262305  0.472488
   4 │ 0.0129008  0.990346
   5 │ 0.368682   0.247991
   6 │ 0.0210636  0.296448
   7 │ 0.473996   0.130082
   8 │ 0.680537   0.159793
   9 │ 0.0541043  0.572126
  10 │ 0.93897    0.050982
```

I have to fix the parsing to avoid that `nothing` in there. Will do

pdeffebach replied:
I suppose I should also special-case `AsTable` on the LHS. But I can do that later.

bkamins replied:
OK - thank you!

nalimilan commented on src/macros.jl:
```diff
@@ -630,6 +630,39 @@ macro with(d, body)
+other columns may be referenced in the block.
+
+Using `AsTable` in this way is useful for working with many columns
+at once programatically. For example, to take a row-wise sum of the
```
```suggestion
at once programmatically. For example, to compute the row-wise sum of the
```

nalimilan commented on src/macros.jl:
````diff
@@ -630,6 +630,39 @@ macro with(d, body)
+columns `[:a, :b, :c, :d]`, write
+
+```
+nms = [:a, :b, :c, :d]
+@byrow :c = sum(AsTable(nms))
````
Why not just
```suggestion
@byrow :c = sum(AsTable([:a, :b, :c, :d]))
```

nalimilan commented on src/macros.jl:
````diff
@@ -630,6 +630,39 @@ macro with(d, body)
+@byrow :c = sum(AsTable(nms))
+```
+
+The constructs the pair
````
```suggestion
This constructs the pairs
```

nalimilan commented on src/macros.jl:
```diff
@@ -630,6 +630,39 @@ macro with(d, body)
     esc(with_helper(d, body))
 end
 
+# TODO: New docstring for @subset
```
What does this mean?

pdeffebach replied:
The current docstring only talks about transformations, i.e. creating new columns. But `@subset` and `@orderby` don't do that, and thus need their own docstrings.

pdeffebach replied:
Actually, @nalimilan I could use your advice on what to do here. 

There is a trade-off between consolidating documentation, having too-long of docstrings, and having all the information in the docstring. 

Should I write variants of this text for every macro? Or put this in the documentation `.md` file separately?

bkamins replied:
My opinion:
* Docstring should be a complete contract for a given functionality.
* If some functionality is shared you can give a reference to a different docstring.
* Explanations of why and how should be given in a manual.

pdeffebach replied:
Okay I've added a pretty lengthy discussion in the `index.md` file. 

@nalimilan @bkamins this is ready for another review.

nalimilan reviewed (commented):
Just a few stylistic comments.

pdeffebach commented on docs/src/index.md:
```diff
@@ -525,11 +525,111 @@ To reference columns with more complicated expressions, you must wrap column ref
+
+| Operation         | Purpose                                                                             | Example           | Notes |
+|-------------------|-------------------------------------------------------------------------------------|-------------------|-------|
+| `$AsTable` on LHS | Create multiple columns at once, whose column names are only known programmatically | <pre lang=julia>$AsTable = f(:y)</pre> | Requires escaping with `$` until deprecation period ends for unquoted column names on LHS.      |
```
Pretty hacky way to make a reasonable little table. But it looks nice when compiled.

pdeffebach replied:
It does not work, unfortunately. I will think of another way to do this.

bkamins commented on docs/src/index.md:
````diff
@@ -525,11 +525,111 @@ To reference columns with more complicated expressions, you must wrap column ref
+To make a new column which is the sum of `vars`, write
+
+```
+julia> @rtransform df :y = sum(AsTable(vars))
````
maybe also add an example when just a column selector is passed to `AsTable` as it will be maybe more common?

pdeffebach replied:
Added. And another example about how function can now "see" the names of columns
.

bkamins commented on docs/src/index.md:
```diff
@@ -525,11 +525,111 @@ To reference columns with more complicated expressions, you must wrap column ref
+If an argument is entirely wrapped in `$()`, the result bypasses the anonymous function 
+creation of DataFramesMeta.jl and is passed to the underling DataFrames.jl function 
+directly. Importantly, this allows for `src => fun => dest` calls from the DataFrames.jl 
+"mini-language" directly.
```
again - maybe add an example? something like `[:a, :b] .=> [sum mean]`? as I guess this is hard to express in DataFramesMeta.jl?

nalimilan commented on docs/src/index.md:
```diff
@@ -525,11 +525,111 @@ To reference columns with more complicated expressions, you must wrap column ref
+
+At this point we have seen `AsTable` appear in three places:
+
+1. `AsTable` on the left-hand-side of transformations: `$AsTable = f(:a, :b)`
```
```suggestion
1. `AsTable` on the left-hand side of transformations: `$AsTable = f(:a, :b)`
```

pdeffebach commented:
Thanks for the review! 

I have addressed all comments. I also got rid of the code example in the table because I couldn't figure out a way to get it to work. 

I think this is ready to merge?

bkamins commented on docs/src/index.md:
```diff
@@ -525,11 +525,165 @@ To reference columns with more complicated expressions, you must wrap column ref
+In operations, it is also allowed to use `AsTable(cols)` to work with
+multiple columns at once, where the columns are grouped together in a
+`NamedTuple`. When `AsTable(cols)` appears in a operation, no
+other columns may be referenced in the block.
```
and it can appear only once - right?

pdeffebach replied:
No. it can appear multiple times. But it has to look the exact same, you can't have, for example

```
@transform df :y = sum(AsTable([:a, :b])) + sum(AsTable(["a", "b"]))
```

There are similar restrictions elsewhere, for example with `@astable`.

bkamins commented on docs/src/index.md:
````diff
@@ -525,11 +525,165 @@ To reference columns with more complicated expressions, you must wrap column ref
+
+To make a new column which is the sum of `vars`, write
+
+```
````
this is not for this PR if it is not covered, but do you resolve it to `AsTable(vars) => sum => :y` (so that we use an efficient path in DataFrames.jl)?

bkamins replied:
I guess the answer is yes given your comment below.

bkamins commented on docs/src/index.md:
```diff
@@ -525,11 +525,165 @@ To reference columns with more complicated expressions, you must wrap column ref
-### Multi-argument selectors 
+`AsTable` on the right-hand side also allows the use of the special
+column selectors `Not`, `Between`, and regular expressions, as well
+as working with lists of variables programmatically. 
```
I understand that anything inside `AsTable` is untouched by the macro - right? Maybe add this comment.

bkamins replied:
OK - I see you answer this question below.

bkamins commented on docs/src/index.md:
````diff
@@ -525,11 +525,165 @@ To reference columns with more complicated expressions, you must wrap column ref
+```
+julia> function fun_with_new_name(x::NamedTuple)
+           nms = string.(propertynames(x))
+           new_name = join(nms, "_") * "_sum"
````
```suggestion
           new_name = Symbol(join(nms, "_"), "_sum")
```

bkamins commented on docs/src/index.md:
```diff
@@ -525,11 +525,165 @@ To reference columns with more complicated expressions, you must wrap column ref
+julia> function fun_with_new_name(x::NamedTuple)
+           nms = string.(propertynames(x))
+           new_name = join(nms, "_") * "_sum"
+           s = sum(x)
+           (; Symbol(new_name) => s)
```
```suggestion
           s = sum(x)
           (; new_name => s)
```

bkamins commented on docs/src/index.md:
````diff
@@ -525,11 +525,165 @@ To reference columns with more complicated expressions, you must wrap column ref
+For example
+
+```
+:y = first(AsTable("a"))
````
why is this an example of the behavior you describe? I guess you wanted `:a` not `"a"` - right?

pdeffebach replied:
No this is correct. Usually when working with `"a"`, you would need to write `$"a"`. I guess I could make it a variable you have to interpolate.

bkamins commented on docs/src/index.md:
```diff
@@ -525,11 +525,165 @@ To reference columns with more complicated expressions, you must wrap column ref
+─────┼──────────────────────────────────────────────
+   1 │     1     30      3     70      1.5     35.0
+   2 │     2     40      3     70      1.5     35.0
+
```
```suggestion
```

bkamins commented on src/macros.jl:
```diff
@@ -349,7 +349,7 @@ macro passmissing(args...)
 end
 
-const astable_docstring_snippet = """
+const astable_macro_flag_docs = """
```
maybe uppercase the constant name?

bkamins commented on src/macros.jl:
```diff
@@ -625,12 +625,102 @@ julia> @with df @byrow :x * :y
     If the parent scope is a local scope (inside a function or let block for example),
     the `global` keyword is not needed to assign to that parent scope.
+
+!!! note
```
should this be indented by 4 spaces?

pdeffebach replied:
Nope it works

pdeffebach replied:
Though I may need you to write the implementation for this in the future since I will have to rely on DataFrames internals to really get this working.

bkamins replied:
So why do you indent `!!! note` above (line 621) with 4 spaces and not here?

pdeffebach replied:
I see for spaces. Can you add a change via github? (We can also worry about this in another PR)

bkamins replied:
ah - you are right. I have mistakenly mixed this with line 352.

bkamins commented on src/macros.jl:
```diff
@@ -625,12 +625,102 @@ julia> @with df @byrow :x * :y
     esc(with_helper(d, body))
 end
 
+astable_rhs_orderby_docs = """
```
make it a `const` and uppercase the name?

pdeffebach replied:
added.

bkamins commented on src/macros.jl:
````diff
@@ -625,12 +625,102 @@ julia> @with df @byrow :x * :y
+This constructs the pair
+
+```
+AsTable(AsTable([:a, :b, :c])) => ByRow(sum)
````
probably a single `AsTable` should be here?

bkamins commented on src/macros.jl:
````diff
@@ -625,12 +625,102 @@ julia> @with df @byrow :x * :y
+```
+"""
+
+astable_rhs_subset_docs = """
````
make it a `const` and uppercase.

bkamins commented on src/macros.jl:
````diff
@@ -625,12 +625,102 @@ julia> @with df @byrow :x * :y
+This constructs the pair
+
+```
+AsTable([:a, :b, :c]) => ByRow(t -> sum(t) > 5)
````
out of curiosity - and for the future - we  could think what syntax could be used in such cases to create:
```
AsTable([:a, :b, :c]) => ByRow((t -> sum(t) > 5)∘collect)
```

pdeffebach replied:
I doubt it. No way of knowing how the function uses the named tuple and I don't want to special-case the `sum` as a syntax.

bkamins commented on src/macros.jl:
````diff
@@ -625,12 +625,102 @@ julia> @with df @byrow :x * :y
+```
+"""
+
+astable_rhs_select_transform_docs = """
````
make it a `const` and uppercase

pdeffebach replied:
added.

bkamins commented:
Looks good. I made some cosmetic comments.

nalimilan commented on docs/src/index.md:
```diff
@@ -627,7 +660,28 @@ The differences between the three is summarized below
 creation of DataFramesMeta.jl and is passed to the underling DataFrames.jl function 
 directly. Importantly, this allows for `src => fun => dest` calls from the DataFrames.jl 
-"mini-language" directly.
+"mini-language" directly. One example where this is useful is calling multiple functions across multiple input parameters. For insteance, the `Pair`
```
```suggestion
"mini-language" directly. One example where this is useful is calling multiple functions across multiple input parameters. For instance, the `Pair`
```

nalimilan commented on docs/src/index.md:
````diff
@@ -627,7 +660,28 @@ The differences between the three is summarized below
+[:a, :b] .=> [sum mean]
+```
+
+takes the `sum` and `mean` of both columns `:a` and `:b` separately. It is not possible to express this with DataFrames.jl. But the operation can easily be performed with `$`
````
Is this what you meant?
```suggestion
takes the `sum` and `mean` of both columns `:a` and `:b` separately. It is not possible to express this with DataFramesMeta.jl. But the operation can easily be performed with `$`
```

nalimilan commented on docs/src/index.md:
````diff
@@ -525,11 +525,165 @@ To reference columns with more complicated expressions, you must wrap column ref
 ```
 
-## Constructing multi-column arguments and `src => fun => dest` calls using `$`
+## Operations with multiple columns at once using `AsTable` inside operations
````
Again, why not put this section right after the one about `@astable`?

pdeffebach replied:
Thanks. Re-arranged.

pdeffebach reviewed (commented):
Thanks for the review. I addressed all comments.

pdeffebach reviewed (commented):
Okay ready for review / merging.

bkamins commented:
I went through the comments and resolved all that I thought were addressed. I left one related to indentation (a minor issue) + one regarding the order of presentation of things (so that @nalimilan can check if now we have what he wants).

Thank you for working on thins!

bkamins commented on src/macros.jl:
```diff
@@ -625,12 +625,102 @@ julia> @with df @byrow :x * :y
     If the parent scope is a local scope (inside a function or let block for example),
     the `global` keyword is not needed to assign to that parent scope.
+
+!!! note
+    Using `AsTable` inside `@with` block is currently not supported.
```
```suggestion
    !!! note
        Using `AsTable` inside `@with` block is currently not supported.
```

pdeffebach commented:
Awesome! Thanks for all the hard work reviewing this.
